frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.tsx

```diff
@@ -0,0 +1,121 @@
+import { ElementType } from '~/types'
+import { SelectableElement } from './SelectableElement'
+import { AlertMessage } from 'lib/lemon-ui/AlertMessage'
+import { htmlElementsDisplayLogic } from 'lib/components/HtmlElementsDisplay/htmlElementsDisplayLogic'
+import { useActions, useValues } from 'kea'
+import { useState } from 'react'
+import { CodeSnippet } from 'lib/components/CodeSnippet'
+
+function indent(level: number): string {
+    return Array(level).fill('    ').join('')
+}
+
+function CloseAllTags({ elements }: { elements: ElementType[] }): JSX.Element {
+    return (
+        <>
+            {[...elements]
+                .reverse()
+                .slice(1)
+                .map((element, index) => (
+                    <pre
+                        className="whitespace-pre-wrap break-all bg-default-dark p-0 m-0 rounded-none text-white text-sm"
+                        key={index}
+                    >
+                        {indent(elements.length - index - 2)}
+                        &lt;/{element.tag_name}&gt;
+                    </pre>
+                ))}
+        </>
+    )
+}
+
+function Tags({
+    elements,
+    highlight,
+    editable,
+    onChange,
+}: {
+    elements: ElementType[]
+    highlight: boolean
+    editable: boolean
+    onChange: (i: number, s: string) => void
+}): JSX.Element {
+    return (
+        <>
+            {elements.map((element, index) => {
+                return (
+                    <SelectableElement
+                        key={`${element.tag_name}-${index}`}
+                        element={element}
+                        isDeepestChild={index === elements.length - 1}
+                        onChange={(s) => (editable ? onChange(index, s) : undefined)}
+                        readonly={!editable}
+                        indent={indent(index)}
+                        highlight={highlight}
+                    />
+                )
+            })}
+        </>
+    )
+}
+
+let uniqueNode = 0
+
+export function HtmlElementsDisplay({
```

Comment:
We usually don't PascalCase abbreviations, so it should be `HTMLElementsDisplay`